compiler/rustc_ast_lowering/src/delegation.rs

```diff
@@ -0,0 +1,317 @@
+use crate::{ImplTraitPosition, ResolverAstLoweringExt};
+
+use super::{ImplTraitContext, LoweringContext, ParamMode};
+
+use ast::visit::Visitor;
+use hir::def::{DefKind, PartialRes, Res};
+use hir::{BodyId, HirId};
+use rustc_ast as ast;
+use rustc_ast::*;
+use rustc_errors::ErrorGuaranteed;
+use rustc_hir as hir;
+use rustc_hir::def_id::DefId;
+use rustc_middle::span_bug;
+use rustc_middle::ty::ResolverAstLowering;
+use rustc_span::{symbol::Ident, Span};
+use rustc_target::spec::abi;
+use std::iter;
+
+// id's naming scheme:
+//
+// item_id - id of delegation item
+// path_id - id of delegation path
+// sig_id - id of item from which the signature is inherited
```

Comment:
Could you give a sketch of what the ast->hir transformation looks like? This file appears to do a lot of non-trivial constructions that could use a summarised form.

Reply:
please do provide an example desugaring here